### PR TITLE
ColumnVector::replicate() is even more readily vectorized now

### PR DESCRIPTION
**Pull request by Akazz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Another minor performance improvement to ColumnVector::replicate() - an even further improvement to #9293

_This is most useful for generating synthetic data in (performance) tests._

Notice the overhead decrease:
```
"Before":

  25.18%  QueryPipelineEx  clickhouse  [.] DB::ColumnVector<unsigned short>::replicate                                                                                                                                                              
  23.18%  QueryPipelineEx  clickhouse  [.] DB::ColumnVector<unsigned int>::replicate                                                                                                                                                                
  19.81%  QueryPipelineEx  clickhouse  [.] DB::(anonymous namespace)::TypedExecutorInvoker<DB::FunctionsLogicalDetail::AndImpl, ... >::apply
  11.98%  QueryPipelineEx  clickhouse  [.] DB::(anonymous namespace)::NumbersSource::generate
  ...

"After":

  33.29%  QueryPipelineEx  clickhouse  [.] DB::(anonymous namespace)::TypedExecutorInvoker<DB::FunctionsLogicalDetail::AndImpl, ... >::apply
  22.21%  QueryPipelineEx  clickhouse  [.] DB::(anonymous namespace)::NumbersSource::generate
   9.35%  QueryPipelineEx  clickhouse  [.] DB::ColumnVector<unsigned int>::replicate
   4.43%  QueryPipelineEx  clickhouse  [.] DB::ColumnVector<unsigned short>::replicate
  ...
```

### CODE REVIEW

**Comment by amosbird** [comment]
It would be ideal to see the changed assembly and with a comment about how this helps vectorization. BTW is it compiler related?

File: dbms/src/Columns/ColumnVector.cpp [hunk 1]
```diff
@@ -339,7 +339,7 @@ ColumnPtr ColumnVector<T>::index(const IColumn & indexes, size_t limit) const
 template <typename T>
 ColumnPtr ColumnVector<T>::replicate(const IColumn::Offsets & offsets) const
 {
-    size_t size = data.size();
+    const size_t size = data.size();
```
Comment (amosbird) [inline dbms/src/Columns/ColumnVector.cpp]: I doubt this is significant.

**Comment by alexey-milovidov** [comment]
@amosbird Yes, it is absolutely unclear how does it help vectorization. Akazz invited us to guess :)

**Comment by alexey-milovidov** [comment]
> Unit tests — fail: 1, passed: 5205 

It's my mistake that was in master.
We can merge this PR after adding a comment...

File: dbms/src/Columns/ColumnVector.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Columns/ColumnVector.cpp]: So the hypothesis is that the compiler is thinking that the `size` variable can alias with `curr` if it's a pointer to the same type? Maybe adding __restrict to curr pointer will also help.

**Comment by alexey-milovidov** [comment]
BTW, I don't see the relevant changes in performance test. We can merge anyway.

File: dbms/src/Columns/ColumnVector.cpp (hunk 1 above)
Comment (Akazz) [reply dbms/src/Columns/ColumnVector.cpp]: This change was meant to be solely cosmetic! And I apologize for the obvious confusion that it has introduced

**Comment by Akazz** [comment]
First off, here's the request that I am testing against:
```
SELECT count() FROM (SELECT materialize(toUInt16(1)) AS x1, materialize(toUInt32(1)) AS x2 FROM numbers(20000000000)) WHERE NOT ignore(and(x1,x2))
```
I found no significant difference in performance depending on compiler (GCC 9.2.1 vs Clang 9.0.0), although the generated code can be quite different.

Here is a snippet of disassembly as reported by perf (compiled with GCC 9.2.1):
```
       │     for (size_t i = 0; i < size; ++i)                                                                                                                      
       │     {                                                                                                                                                      
       │     const auto span_end = res->getData().begin() + offsets[i];                                                                                             
       │     for (; it < span_end; ++it)                                                                                                                            
       │       test    %rdx,%rdx                                                                                                                                    
       │     → jle     9031114 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&) const+0x1
       │       mov     0x10(%rbp),%rsi                                                                                                                              
       │       add     %rdi,%rsi                                                                                                                                    
       │       nop                                                                                                                                                  
       │     *it = data[i];                                                                                                                                         
       │       mov     (%rsi),%edx                                                                                                                                  
       │     void increment() { ++m_iterator; }                                                                                                                     
 33.18 │       add     $0x4,%rax                                                                                                                                    
       │       mov     %edx,-0x4(%rax)                                                                                                                              
       │     return y.base() - m_iterator;                                                                                                                          
 31.52 │       mov     %rcx,%rdx                                                                                                                                    
       │       sub     %rax,%rdx                                                                                                                                    
       │     for (; it < span_end; ++it)                                                                                                                            
 35.05 │       test    %rdx,%rdx                                                                                                                                    
       │     → jg      9031100 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&) const+0x1
       │     for (size_t i = 0; i < size; ++i)                                                                                                                      
       │       add     $0x4,%rdi                                                                                                                                    
       │       cmp     %rbx,%rdi                                                                                                                                    
       │     → jne     90310e0 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&) const+0x1
```
Unfortunately, compiler is unable to vectorize the inner loop above.

The proposed patch makes it right.
GCC 9.2.1:
```
       │                 *curr++ = data[i];
 23.74 │1a0:┌─→movups %xmm0,(%rax)
       │    │        while (curr < span_end)
 66.17 │    │  add    $0x10,%rax
  8.05 │    ├──cmp    %rsi,%rax
       │    └──jne    902cf00 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1a0
```

Clang 9.0.0:
```
       │     ↓ je     8a6cd52 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1a2
       │       movd   (%rcx),%xmm0
       │       pshufd $0x0,%xmm0,%xmm0
       │       mov    %r15,%r11
       │       sub    %rax,%r11
       │       xor    %eax,%eax
       │       nop
       │       nop
  1.69 │140:┌─→movdqu %xmm0,(%rdi,%rax,4)
 37.23 │    │  movdqu %xmm0,0x10(%rdi,%rax,4)
 26.62 │    │  movdqu %xmm0,0x20(%rdi,%rax,4)
 13.69 │    │  movdqu %xmm0,0x30(%rdi,%rax,4)
 19.85 │    │  add    $0x10,%rax
       │    ├──add    $0x2,%r11
       │    └──jne    8a6ccf0 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 140
       │       test   %r15,%r15
       │     ↓ je     8a6cd2a <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 17a
       │                 *curr++ = data[i];
       │166:   movd   (%rcx),%xmm0
       │       pshufd $0x0,%xmm0,%xmm0
       │       movdqu %xmm0,(%rdi,%rax,4)
       │       movdqu %xmm0,0x10(%rdi,%rax,4)
       │17a:   lea    (%rdi,%r13,4),%rdi
       │             while (curr < span_end)
       │       cmp    %r13,%r12
       │       mov    -0x70(%rbp),%r15
       │     ↑ je     8a6cc50 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, a0
```

**Comment by Akazz** [comment]
I've looked into it a little bit deeper and isolated the problem to be with ... `<` - the comparison operator! See the new patch!
GCC 9.2.1:
```
       │                 *it = data[i];
 18.65 │1b0:┌─→movups %xmm0,(%rax)
       │    │        for (; it != span_end; ++it)
 70.93 │    │  add    $0x10,%rax
  8.78 │    ├──cmp    %rsi,%rax
       │    └──jne    90317e0 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1b0
  0.18 │       mov    %r9,%rsi
       │       and    $0xfffffffffffffffc,%rsi
  0.18 │       lea    (%rcx,%rsi,4),%rax
       │       cmp    %rsi,%r9
       │     ↓ je     903181c <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1ec
       │                 *it = data[i];
       │       mov    (%r8),%ecx
       │           }
       │
       │           void increment() { ++m_iterator; }
       │       lea    0x4(%rax),%rsi
       │       mov    %ecx,(%rax)
       │             for (; it != span_end; ++it)
       │       cmp    %rsi,%rdi
       │     ↓ je     903181c <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1ec
       │                 *it = data[i];
       │       mov    %ecx,0x4(%rax)
       │       lea    0x8(%rax),%rcx
       │             for (; it != span_end; ++it)
       │       cmp    %rcx,%rdi
       │     ↓ je     903181c <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 1ec
       │                 *it = data[i];
       │       mov    (%r8),%ecx
       │       mov    %ecx,0x8(%rax)
  0.18 │1ec:   mov    %r10,%rcx
       │         for (size_t i = 0; i < size; ++i)
  0.18 │1ef:   cmp    %rdx,%rbx
       │     ↑ jne    9031760 <DB::ColumnVector<unsigned int>::replicate(DB::PODArray<unsigned long, 4096ul, 130
```

**Comment by danlark1** [comment]
The problem is in boost::iterator https://gcc.godbolt.org/z/qgwFCF

**Comment by alexey-milovidov** [comment]
It's clearly better.
```
SELECT count() FROM (SELECT materialize(1) AS x1, materialize(1) AS x2, materialize(1) AS x3, materialize(1) AS x4, materialize(1) AS x5, materialize(1) AS x6, materialize(1) AS x7, materialize(1) AS x8, materialize(1) AS x9, materialize(1) AS x10 FROM numbers(500000000)) WHERE NOT ignore(xor(x1,x2,x3,x4,x5,x6,x7,x8,x9,x10))
```
- 39% faster!

**Comment by alexey-milovidov** [comment]
@danlark1 Thank you! I will try to get rid of boost::iterator.

**Pull request merged**